File: 14.longest-common-prefix/python/src/longest_common_prefix.py
class Solution:
    def longest_common_prefix(self, strs: list[str]) -> str:
        if (not self.is_valid(strs)):
            return ""
        
        low = 1
        high = self.get_min_length(strs)

        while(low <= high):
            mid = (low + high) // 2
            if self.is_common_prefix(strs, mid):
                low = mid + 1
            else:
                high = mid - 1

        result = (low + high) // 2
        return strs[0][0:result]
        
    def is_valid(self, strs: list[str]) -> bool:
        if (strs == None or len(strs) == 0):
            return False

        return True

    def get_min_length(self, strs: list[str]) -> int:
        minStr = min(strs, key=len)
        return len(minStr)

    def is_common_prefix(self, strs: list[str], mid: int) -> bool:
        prefix = strs[0][0:mid]
        for i in range(1, len(strs)):
            if not strs[i].startswith(prefix):
                return False

        return True

File: 14.longest-common-prefix/nodejs/src/longest_common_prefix.ts
class Solution {
    public longestCommonPrefix(strs: string[]): string {
        if (! this.isValid(strs)) {
            return ""
        }

        let low: number = 1
        let high: number = this.getMinLen(strs)
        let mid: number

        while (low <= high) {
            mid = this.toInt((low + high) / 2)

            if (this.isCommonPrefix(mid, strs)) {
                low = mid + 1
                continue
            }

            high = mid - 1
        }

        const result = this.toInt((low + high) / 2)
        return strs[0].slice(0, result)
    }

    private isValid(strs: string[]): boolean {
        return ! (strs === undefined || strs === null || strs.length === 0)
    }

    private getMinLen(strs: string[]): number {
        let minLen: number = strs[0].length

        for (let i = 1; i < strs.length; i++) {
            if (strs[i].length < minLen) {
                minLen = strs[i].length
            }
        }

        return minLen
    }

    private toInt(num: number): number {
        return Math.trunc(num)
    }

    private isCommonPrefix(mid: number, strs: string[]): boolean {
        const prefix: string = strs[0].slice(0, mid)

        for (let i = 1; i < strs.length; i++) {
            if (! (strs[i].startsWith(prefix))) {
                return false
            }
        }

        return true
    }
}

export { Solution }
